pymatgen/io/lobster/outputs.py

```diff
@@ -1363,24 +1365,31 @@ def _read(self, contents: list, spin_numbers: list):
                 kpoint_array = []
                 for kpointel in kpoint:
                     if kpointel not in ["at", "k-point", ""]:
-                        kpoint_array.append(str(kpointel))
+                        kpoint_array += [float(kpointel)]
 
             elif "maxDeviation" in line:
                 if spin not in self.band_overlaps_dict:
                     self.band_overlaps_dict[spin] = {}
-                if " ".join(kpoint_array) not in self.band_overlaps_dict[spin]:
-                    self.band_overlaps_dict[spin][" ".join(kpoint_array)] = {}
+                if "k_points" not in self.band_overlaps_dict[spin]:
+                    self.band_overlaps_dict[spin]["k_points"] = []
+                if "max_deviations" not in self.band_overlaps_dict[spin]:
+                    self.band_overlaps_dict[spin]["max_deviations"] = []
+                if "matrices" not in self.band_overlaps_dict[spin]:
+                    self.band_overlaps_dict[spin]["matrices"] = []
                 maxdev = line.split(" ")[2]
-                self.band_overlaps_dict[spin][" ".join(kpoint_array)]["maxDeviation"] = float(maxdev)
-                self.max_deviation.append(float(maxdev))
-                self.band_overlaps_dict[spin][" ".join(kpoint_array)]["matrix"] = []
+                self.band_overlaps_dict[spin]["max_deviations"] += [float(maxdev)]
+                self.band_overlaps_dict[spin]["k_points"] += [kpoint_array]
+                self.max_deviation += [float(maxdev)]
+                overlaps = []
 
             else:
-                overlaps = []
+                rows = []
                 for el in line.split(" "):
                     if el != "":
-                        overlaps.append(float(el))
-                self.band_overlaps_dict[spin][" ".join(kpoint_array)]["matrix"].append(overlaps)
+                        rows += [float(el)]
+                overlaps += [rows]
```

Comment:
`overlaps` is possibly unbound now from renaming `overlaps = []` to `rows` above.

we don't check for this in CI yet due to way too many legacy violations but really should. related discussion in https://github.com/materialsproject/pymatgen/pull/3646

Reply:
Hi @janosh, I did not get what you meant here. Can you please elaborate a bit more? I need the overlaps variable initialized in the previous if clause to properly store matrices for each k-point.

Is there anything I need to do ? I could not think of any other way to do this.

Reply:
the problem is, if the code enters the `else` case `overlaps` won't be declared, so Python will throw a `NameError`

Reply:
I think, such case won't happen as per the current file format, it will always pass through the [elif clause ](https://github.com/naik-aakash/pymatgen/blob/b93fc2331a2a5f8e69483f318762599208642ab4/pymatgen/io/lobster/outputs.py#L1370).

But I Will check for a few more examples that I have and see If I encounter any errors

Reply:
Hi @janosh , I checked for about 20 different example files which I have, does not result in `NameError `due to file format , logic seems to work without breaking it 😃

But if you have any better idea to get same outcome, I am happy to implement it. Just at this point I seem to have blanked out 😅 and cannot think of any alternatives

Reply:
you could `pip install pyright` and run it on this file. it'll show you the error i mean. you just need to ensure `overlaps` is declared by all code paths

Reply:
Hi @janosh , I have now address the errors which I could using pyright. If you think these are okay it could be merged.